pci/src/vfio.rs

```diff
@@ -376,6 +376,19 @@ impl VfioCommon {
             let region_size: u64;
             let bar_addr: GuestAddress;
 
+            let base_addr = if let Some(resources) = &resources {
+                if resources.len() > ranges.len() {
+                    return Err(PciDeviceError::MissingResource);
+                }
+                match resources[ranges.len()] {
+                    Resource::MmioAddressRange { base, .. } => Some(GuestAddress(base)),
+                    Resource::PioAddressRange { base, .. } => Some(GuestAddress(base as u64)),
```

Comment:
Related to my other question: here we can potentially "restore" from a (old) resources with `Resource::PioAddressRange` entries, while we always return a "new_resource" with only `Resource::MmioAddressRange` entries from `add_pci_device()` (who is the caller of this function `move_bars()`). They seem to be conflicting, right? I may lack some general background here.

Reply:
Good catch, I must create the `new_resource` more carefully, checking the type of each BAR.

Reply:
But this will be done part of the commit where I factorize into `add_pci_device()` because so far it's correct for virtio-pci devices as we know we don't create PIO BARs for them.

Reply:
Could use use  `impl From<..>` etc to remove duplication.

Reply:
Well if you mean something like `impl From<(GuestAddress, u64, PciBarRegion)> for Resource`, that's complicating things quite a lot because `Resource` comes from `vm-device`, which is imported from `pci`. But `PciBarRegion` is defined in `pci`, meaning we run into a cyclic dependency.
I think we should leave the very small amount of duplication for now and maybe work towards a better approach in the future as I mentioned [here](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3952#discussion_r849579006).
We could even stop using a tuple to refer to the BARs and instead rely on the newly created `Resource::Bar` to carry the information we need. This way we wouldn't have to translate the types.